sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+    for tid, (attack, episode_subsequence) in enumerate(episode_subsequences.items()):
 
-        if len(episodes) < 3:
+        if len(episode_subsequence) < 3:
             continue
-        #print(' ------------- COUNTER ', counter, '------')
         counter += 1
         
-        if '10.0.254' not in attacker:
+        if '10.0.254' not in attack:  # TODO: this part has to be commented for CCDC dataset
             continue
-        if ('147.75' in attacker or '69.172'  in attacker):
-                continue
-        tr = [int(x) for x in state_traces[counter]]
-        #print(counter)
-        num_servs = [len(set((x[6]))) for x in episodes]
-        max_servs = [most_frequent(x[6]) for x in episodes]
-        #print(max_servs)
-        
-        if 0 in tr and (not set(tr).isdisjoint(sev_states) or not set(tr).isdisjoint(med_states)):
-            levelone.add(tr[tr.index(0)+1])
-       
-        #print([x[2] for x in episodes]) 
-        #print(state_traces[counter])
-        new_state = (state_traces[counter][1:])[::-1]
+        if '147.75' in attack or '69.172' in attack:
+            continue
+
+        trace = [int(state) for state in state_traces[counter]]
+        max_services = [_most_frequent(epi[6]) for epi in episode_subsequence]
+
+        if 0 in trace and (not set(trace).isdisjoint(sev_states) or not set(trace).isdisjoint(med_states)):
+            level_one.add(trace[1])
+
+        trace = trace[1:][::-1]  # Reverse the trace from the S-PDFA back
         
-        #print(new_state, [x[2] for x in episodes])
-        # also artifically add tiny delay so all events are not exactly at the same time.
-        #print(len(episodes), new_state, max_servs)
-        times = [(x[0], x[1], x[2], int(new_state[i]), max_servs[i], x[7], x[8]) for i,x in enumerate(episodes)] # start time, endtime, episode mas, state ID, most frequent service, list of unique signatures, (1st and last timestamp)
+        # start_time, end_time, mcat, state_ID, mserv, list of unique signatures, (1st and last timestamp)
+        state_subsequence = [(epi[0], epi[1], epi[2], trace[i], max_services[i], epi[7], epi[8])
+                                for i, epi in enumerate(episode_subsequence)]
         
-        step1 = attacker.split('->')
-        step1_0 = step1[0].split('-')[0]
-        step1_1 = step1[0].split('-')[1]
-        step2 = step1[-1].split('-')[0]
-        real_attacker = '->'.join([step1_0+'-'+step1_1, step2])
-        real_attacker_inv = '->'.join([step1_0+'-'+step2, step1_1])
-        #print(real_attacker)
-        INV = False
-        if '10.0.254' in step2:
-            INV = True
+        parts = attack.split('->')
+        team, attacker = parts[0].split('-')
+        victim, attack_num = parts[1].split('-')
+        attacker_victim = team + '-' + attacker + '->' + victim
+        attacker_victim_inv = team + '-' + victim + '->' + attacker
+        inv = False
+        if '10.0.254' in victim:
```

Comment:
I think this is CPTC-specific, but I am also not sure what to do with this, so for this PR it can be here. I think the reversing happens only here, i.e. after FlexFringe...

Reply:
yeah, we probably need to do this as soon as the alerts are parsed and then keep the order. This comes way too late